rclrs/src/node/primitive_options.rs

```diff
@@ -0,0 +1,253 @@
+use crate::{
+    QoSDurabilityPolicy, QoSDuration, QoSHistoryPolicy, QoSLivelinessPolicy, QoSProfile,
+    QoSReliabilityPolicy,
+};
+
+use std::{borrow::Borrow, time::Duration};
+
+/// `PrimitiveOptions` are the subset of options that are relevant across all
+/// primitives (e.g. [`Subscription`][1], [`Publisher`][2], [`Client`][3], and
+/// [`Service`][4]).
+///
+/// Each different primitive type may have its own defaults for the overall
+/// quality of service settings, and we cannot know what the default will be
+/// until the `PrimitiveOptions` gets converted into the more specific set of
+/// options. Therefore we store each quality of service field separately so that
+/// we will only override the settings that the user explicitly asked for, and
+/// the rest will be determined by the default settings for each primitive.
+///
+/// [1]: crate::Subscription
+/// [2]: crate::Publisher
+/// [3]: crate::Client
+/// [4]: crate::Service
+#[derive(Debug, Clone, Copy)]
+#[non_exhaustive]
+pub struct PrimitiveOptions<'a> {
+    /// The name that will be used for the primitive
+    pub name: &'a str,
```

Comment:
This lifetime is rather annoying, it kind of leaks everywhere. Perhaps its acceptable to own this data w/ `String`? I guess the downside is we'd need to resolve this if we ever wanted something like `no_std` support. Also I am assuming no one would ever create this in a "hot loop" so the allocation shouldn't be too much overhead...

Reply:
Two thoughts:
* We shouldn't have a reflex to shy away from lifetimes, especially when it comes to this `_Options` pattern where we have a structure that is specifically meant to have a temporary existence. For now it's "only" a string allocation, but as we add more options in the future we may have a need to borrow other larger data that could require heap allocations. As far as lifetimes go, this is a pretty clean and easy use of them.
* We're already using this pattern for `ToLogParams` which does need to run in hot loops and benefits enormously from the explicit lifetime. I think it's better to use explicit lifetimes consistently across all our uses of the `_Options` pattern rather than picking and choosing where allocations will be "costly" enough to justify it.

Reply:
> There are 42 cases of this lifetime in this file

In fairness that's only because the name `PrimitiveOptions<'a>` is present 42 times. If you think of the lifetime as just part of the name of `PrimitiveOptions` then the lifetime isn't imposing itself on anything besides the name of this particular struct.

> In fact, if we enforce `'static` lifetime of this str couldn't we elide the lifetime?

We would not want this. This would be fine if users **_only ever_** use compile-time string literals for their topic names, but if a topic name is ever generated at runtime or delivered by a `String` then they would have to [intentionally leak memory](https://doc.rust-lang.org/std/string/struct.String.html#method.leak) in order to get a `&'static str` borrow of it. If we absolutely need to get rid of the lifetime, then changing the `name: &'a str` to `name: String` would be the way to go.

> lets just make sure this annoyance provides substantial downstream benefit.

Can you elaborate on what you perceive the annoyance to be? Maybe we have a misunderstanding about what the lifetime entails. Downstream users of rclrs don't have to spend a single moment thinking about the lifetime. Here are some example snippets of code that would use the API as it currently is:

```rust
// Using string literals
let publisher = node.create_publisher("my_topic")?;
let publisher = node.create_publisher(
    "my_topic"
    .reliable()
    .keep_last(100)
)?;

// Using a heap-allocated string
let my_topic = String::from("my_topic");
let publisher = node.create_publisher(&my_topic)?;
let publisher = node.create_publisher(
    my_topic
    .reliable()
    .keep_last(100)
)?;
```

Users never need to touch or think about the `'a` lifetime. Its existence only affects us as maintainers, and as maintainers its only effect is that we need to carry it along wherever the struct name `PrimitiveOptions` gets used.

---

Admittedly I could see an argument for users wanting to save and distribute one or more instances of `PrimitiveOptions`, e.g. decide what all your topic/service settings should be in one part of your code, save those decisions as `PrimitiveOptions`, and then load that in another part of your code. For that we could consider 

```rust
pub struct PrimitiveOptions<'a> {
    pub name: Cow<'a, str>,
    ...
}
```

to allow users to choose between an owned `String` versus a temporary borrow `&'a str`. Or we could take it a step further to get rid of the lifetime while allowing users to choose between an owned `String` or a static string literal:

```rust
pub struct PrimitiveOptions {
    pub name: Cow<'static, str>,
    ...
}
```

But I think this would be a strange and arbitrary limitation to impose just so we can avoid typing `<'a>` next to `PrimitiveOptions` inside the library source code. I'm also skeptical that there's much value in saving and distributing instances of `PrimitiveOptions` in this fashion. I think it's more likely that users will want a map like `HashMap<String, QoSProfile>` which they would then use to instantiate the primitives like

```rust
for (publisher_name, qos) in publisher_settings {
    let publisher = node.create_publisher(publisher_name.qos(qos))?;
    all_publishers.push(publisher);
}
```
Ultimately end-users of rclrs aren't really meant to use `PrimitiveOptions` directly. They don't even need to know it exists. We only have it to trick the Rust compiler into letting us overload the first argument in each of our `create_[primitive]` functions.